### PR TITLE
RaR: Mind's Eye and Divide and Conquer; do-access refactor and Shiro/Mwanza bugfixes

### PR DESCRIPTION
**Pull request by presheaf**
This is a bit of a doozy. I have implemented Mind's Eye, which basically boils down to making `do-access` usable outside of runs. This was done using Noah's work on D&C. I anticipate there are lots of things which need changing, not least of which is a bug with Shiro/Mwanza, but I figured it would be best to get some feedback on what's been done so far before I make an even larger mess.

### CODE REVIEW

**Comment by presheaf** [comment]
I'm not able to reproduce the failing test locally - here everything seems to work. Could someone else see if they get a failing test too?

File: src/clj/game/cards/upgrades.clj [hunk 1]
```diff
@@ -609,27 +609,28 @@
...
-                                                (str "gains " total " [Credits] from Mwanza City Grid"))))}]
-     {:install-req (req (filter #{"HQ" "R&D"} targets))
-      :events {:pre-access {:req (req (and installed
-                                           ;; Pre-access server is same server as that Mwanza is in the root of
-                                           (= target (second  (:zone card)))))
+   (let [gain-creds-and-clear {:req (req (= (:from-server target) (second (:zone card))))
+                               :silent (req true)
+                               :effect (req (let [cnt (:cards-accessed run)]
+                                              (access-bonus state :runner -3)
+                                              (if cnt ; TODO: this fails if the run featured multiple access phases (e.g. shiro sub firing)
```
Comment (NoahTheDuke) [inline src/clj/game/cards/upgrades.clj]: This needs to be a `when`. Right now, if `cnt` *is* defined, it gains the credits, and if it *isn't* defined, it calls `system-msg`. Change it to a when to only execute both line if `cnt` is defined.

File: src/clj/game/core/runs.clj [hunk 2]
```diff
@@ -648,14 +652,17 @@
...
-(defmethod choose-access :archives [cards server]
+(defmethod choose-access :archives [cards server {:keys [no-root] :as args}]
   {:async true
    :effect (req (let [cards (concat (get-archives-accessible state) (-> @state :corp :servers :archives :content))
                       archives-count (+ (count (-> @state :corp :discard)) (count (-> @state :corp :servers :archives :content)))]
                   (if (not-empty cards)
                     (if (= 1 archives-count)
-                      (access-card state side eid (first cards))
-                      (continue-ability state side (access-helper-archives state archives-count #{}) card nil))
+                      (wait-for (access-card state side eid (first cards)))
```
Comment (NoahTheDuke) [inline src/clj/game/core/runs.clj]: If you're trying to make the line below wait for `access-card` to finish, you have to encompass the following lines in the `wait-for`'s parenthesis.

File: src/clj/game/core/runs.clj [hunk 3]
```diff
@@ -849,33 +861,39 @@
...
-    (swap! state assoc :run nil)
-    (update-all-ice state side)
-    (swap! state dissoc :access)
-    (clear-run-register! state)
-    (trigger-run-end-events state side eid run)))
+      (if-let [end-run-effect (:end-run run-effect)]
+        (wait-for (resolve-ability state side end-run-effect (:card run-effect) [server])
+                  (run-cleanup-2 state side run))
+        (run-cleanup-2 state side run)
+))))
```
Comment (NoahTheDuke) [inline src/clj/game/core/runs.clj]: ERROR: PARENTHESES ALONE. MALFUNCTION IMMINENT.

File: src/clj/game/core/runs.clj (hunk 2 above)
Comment (nealterrell) [reply src/clj/game/core/runs.clj]: You also don't pass your own `eid` to another function that you are awaiting, so remove that parameter.

Comment (NoahTheDuke) [reply src/clj/game/core/runs.clj]: I see now that the `continue-ability` is for the false if-branch, but removing the `wait-for` breaks things, so I don't actually know what's going on here.

File: src/clj/game/cards/events.clj [hunk 4]
```diff
@@ -501,6 +501,20 @@
...
+   "Divide and Conquer"
+   {:req (req archives-runnable)
+    :makes-run true
+    :async true
+    :effect
+    (effect (run :archives
+                 {:end-run
+                  {:async true
+                   :effect
+                   (req (wait-for (do-access state side [:rd] {:no-root true})
```
Comment (Saintis) [inline src/clj/game/cards/events.clj]: Do we have confirmation that the R&D access benefits from additional accesses?

File: src/clj/game/core/runs.clj [hunk 5]
```diff
@@ -687,22 +699,27 @@
...
-                     (system-msg state side "accessed no cards during the run")
-                     (do (swap! state assoc-in [:runner :register :accessed-cards] true)
-                         (wait-for (resolve-ability state side (choose-access cards server) nil nil)
-                                   (effect-completed state side eid))
-                         (swap! state update-in [:run :cards-accessed] (fnil #(+ % n) 0)))))
-                 (handle-end-run state side)))))
+             (let [cards (cards-to-access state side server)
+                   cards (if hq-root-only (remove #(= '[:hand] (:zone %)) cards) cards)
+                   cards (if no-root (remove #(or (= '[:servers :rd :content] (:zone %))
+                                                  (= '[:servers :hq :content] (:zone %))) cards) cards)
```
Comment (Saintis) [inline src/clj/game/core/runs.clj]: Are the quotes needed here? I thought you only had to quote `'()`

File: src/clj/game/core/runs.clj [hunk 6]
```diff
@@ -849,33 +866,39 @@
     :else
     (effect-completed state side eid)))
 
+(defn run-cleanup-2
```
Comment (Saintis) [inline src/clj/game/core/runs.clj]: Needs better name 😛, and a doc-string.

Comment (NoahTheDuke) [reply src/clj/game/core/runs.clj]: Lol Any ideas?

File: src/clj/game/cards/events.clj (hunk 4 above)
Comment (NoahTheDuke) [reply src/clj/game/cards/events.clj]: [Yep.](http://7tr.in/ANR/divide-and-conquer.png) Gonna be in the forthcoming UFAQ.

**Pull request closed without merging**